core/src/server/rpc_module.rs

```diff
@@ -816,98 +846,196 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
+// to be &self instead of &mut self.
+#[derive(Debug)]
+struct IsUnsubscribed(mpsc::Sender<()>);
 
-/// Represents a single subscription.
+impl IsUnsubscribed {
+	/// Returns true if the unsubscribe method has been invoked or the subscription has been canceled.
+	///
+	/// This can be called multiple times as the element in the channel is never
+	/// removed.
+	fn is_unsubscribed(&self) -> bool {
+		self.0.is_closed()
+	}
+
+	/// Wrapper over [`tokio::sync::mpsc::Sender::closed`]
+	///
+	/// # Cancel safety
+	///
+	/// This method is cancel safe. Once the channel is closed,
+	/// it stays closed forever and all future calls to closed will return immediately.
+	async fn unsubscribed(&self) {
+		self.0.closed().await;
+	}
+}
+
+/// Represents a single subscription that is waiting to be accepted or rejected.
+///
+/// If this is dropped without calling `PendingSubscription::reject` or `PendingSubscriptionSink::accept`
+/// a default error is sent out as response to the subscription call.
+///
+/// Thus, if you want a customized error message then `PendingSubscription::reject` must be called.
 #[derive(Debug)]
-pub struct SubscriptionSink {
+#[must_use = "PendningSubscriptionSink does nothing unless `accept` or `reject` is called"]
+pub struct PendingSubscriptionSink {
 	/// Sink.
 	inner: MethodSink,
-	/// Get notified when subscribers leave so we can exit
-	close_notify: Option<SubscriptionPermit>,
 	/// MethodCallback.
 	method: &'static str,
 	/// Shared Mutex of subscriptions for this method.
 	subscribers: Subscribers,
 	/// Unique subscription.
 	uniq_sub: SubscriptionKey,
-	/// Id of the `subscription call` (i.e. not the same as subscription id) which is used
+	/// ID of the `subscription call` (i.e. not the same as subscription id) which is used
 	/// to reply to subscription method call and must only be used once.
-	///
-	/// *Note*: Having some value means the subscription was not accepted or rejected yet.
-	id: Option<(Id<'static>, oneshot::Sender<MethodResponse>)>,
-	/// Having some value means the subscription was accepted.
-	unsubscribe: UnsubscribeCall,
-	/// Claimed resources.
-	_claimed: Option<ResourceGuard>,
+	id: Id<'static>,
+	/// Sender to answer the subscribe call.
+	subscribe: oneshot::Sender<MethodResponse>,
 }
 
-impl SubscriptionSink {
-	/// Reject the subscription call from [`ErrorObject`].
-	pub fn reject(&mut self, err: impl Into<ErrorObjectOwned>) -> Result<(), SubscriptionAcceptRejectError> {
-		let (id, subscribe_call) = self.id.take().ok_or(SubscriptionAcceptRejectError::AlreadyCalled)?;
+impl PendingSubscriptionSink {
+	/// Reject the subscription call with the error from [`ErrorObject`].
+	pub async fn reject(self, err: impl Into<ErrorObjectOwned>) -> Result<(), SubscriptionAcceptRejectError> {
+		let err = MethodResponse::error(self.id, err.into());
+		let permit = self.inner.reserve().await.map_err(|_| SubscriptionAcceptRejectError::RemotePeerAborted)?;
```

Comment:
initially, I had an idea to limit to sends APIs to `MethodSinkPermit` but didn't work because subscriptions may send more than one message.

Thus, you are right those "reserve" are needless.